### ISSUE TITLE
chap06_Keras_Model_tf_Dataset

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_deep_learning#50]
"""
Dataset 클래스
 - dataset으로부터 사용가능한 데이터를 메모리에 로딩 가능
 - batch size 지정
"""

import tensorflow as tf

from tensorflow.python.data import Dataset

- 호출가능한 member 확인
dir(Dataset)
'''
batch() : 비복원 추출이다.
from_tensor_slices()
shuffle()
'''

### 1. from_tensor_slices() : 입력 tensor로부터 slice 생성
ex) MNIST(60000, 28, 28) -> 60000개 image를 각각 1개씩 slice

- 1) x, y변수 생성
x = tf.random.normal([5, 2]) # 2차원
y = tf.random.normal([5]) # 1차원

- 2) dataset : 5개 slice
train_ds = Dataset.from_tensor_slices((x,y))
train_ds # <TensorSliceDataset shapes: ((2,), ()), types: (tf.float32, tf.float32)>

- 5개 관측치 -> 5개 slices
for train_x, train_y in train_ds :
    print("x = {}, y= {}".format(train_x.numpy(), train_y.numpy()))
'''
x = [-1.2234435  1.2301564], y= -1.207342267036438
x = [1.7116114  0.32205266], y= 1.4923869371414185
x = [0.5806598 1.0023462], y= 0.6844696998596191
x = [-1.8977883  2.557473 ], y= 1.916743278503418
x = [ 1.4058231 -1.8076215], y= 1.0934163331985474
'''

### 2. from_tensor_slices(x,y).shuffle(buffer size).batch(size)
'''
shuffle(buffer size) : tensor 행단위로 서플링
- buffer size : 선택된 data size
batch : model에 1회 공급할 dataset size
ex) 60,000(mnist) -> shuffle(10000).batch(100)
    1번째 slice data : 10000개 서플링 -> 순서대로 100개씩(1회 공급 데이터) 추출
    2번째 slice data : next 10000개 가져와서 서플링 -> 100개씩 추출
'''

- 1) x, y변수 생성
x2 = tf.random.normal([5, 2]) # 2차원
y2 = tf.random.normal([5]) # 1차원

- 2) dataset : 5개 관측치 -> 3개 slice
train_ds2 = Dataset.from_tensor_slices((x2,y2)).shuffle(5).batch(2)
train_ds2 # <BatchDataset shapes: ((None, 2), (None,)), types: (tf.float32, tf.float32)>

- 3) 3 slice -> 1 slice
for train_x, train_y in train_ds2 :
    print("x = {}, y= {}".format(train_x.numpy(), train_y.numpy()))
'''
-batch(2) : 2개의 데이터가 모델에 공급할 하나의 데이터가 되는 것.
x = [[-0.32956833 -1.9187714 ]
 [ 0.42682296  0.23169912]], y= [-0.89258313 -0.30307543]
x = [[ 1.1581373  0.7614629]
 [-1.372349   0.9048762]], y= [-1.9779706  0.8967127]
x = [[ 1.3459641 -0.4728109]], y= [0.3576125]
'''

### 3. keras dataset 적용
from tensorflow.keras.datasets.cifar10 import load_data

- 1. dataset load
(x_train, y_train), (x_val, y_val) = load_data()

x_train.shape # images : (50000, 32, 32, 3) - (size, h w,c)
y_train.shape # labels : (50000, 1)

import matplotlib.pyplot as plt

plt.imshow(x_train[0])
plt.show()

y_train[0] #  array([6], dtype=uint8) # frog 개구리

- train set batch size = 100 image
train_ds = Dataset.from_tensor_slices((x_train,y_train)).shuffle(10000).batch(100)
cnt =0
for image_x, label_x in train_ds : 
    print("image = {}, label= {}".format(image_x.shape, label_x.shape))
    cnt += 1
print("slice 개수=", cnt)
epochs = iter size(500) * batch size(100)
총 50000장이 100장씩 공급되므로 슬라이스의 개수는 500개 이다.
'''
image = (100, 32, 32, 3), label= (100, 1)
image = (100, 32, 32, 3), label= (100, 1)
slice 개수= 500
'''

- test set batch size = 100 image
- 총 10000장이 100장씩 공급되므로 슬라이스의 개수느 100개이다.
test_ds = Dataset.from_tensor_slices((x_val,y_val)).shuffle(10000).batch(100)
cnt =0
for image_x, label_x in test_ds : 
    print("image = {}, label= {}".format(image_x.shape, label_x.shape))
    cnt += 1
print("slice 개수=", cnt) 
'''
image = (100, 32, 32, 3), label= (100, 1)
image = (100, 32, 32, 3), label= (100, 1)
image = (100, 32, 32, 3), label= (100, 1)
slice 개수= 100
'''

'''
문) mnist 데이터셋을 이용하여 train_ds, val_ds 생성하기
    train_ds : shuffle = 10,000, batch_size = 32
    val_ds : batch_size = 32
'''
from tensorflow.keras.datasets.mnist import load_data
(x_train, y_train), (x_val, y_val) = load_data()
y_train # label : 10진수

train_ds = Dataset.from_tensor_slices((x_train,y_train)).shuffle(10000).batch(32)
cnt =0
for image_x, label_x in train_ds : 
    print("image = {}, label= {}".format(image_x.shape, label_x.shape))
    cnt += 1
print("slice 개수=", cnt) # slice 개수= 1875
1875 * 32 # 60000

test_ds = Dataset.from_tensor_slices((x_val,y_val)).batch(32)
cnt =0
for image_x, label_x in test_ds : 
    print("image = {}, label= {}".format(image_x.shape, label_x.shape))
    cnt += 1
print("slice 개수=", cnt)  # slice 개수= 313
'''
-만약 데이터 부족하면 앞쪽으로 와서 데이터 사이즈 맞춤
image = (32, 28, 28), label= (32,)
image = (32, 28, 28), label= (32,)
image = (16, 28, 28), label= (16,)
'''
